src/analyses/weaklyRelationalPointerAnalysis.ml

```diff
@@ -0,0 +1,197 @@
+    (* remove callee vars that are not reachable and not global *)
+    let reachable_variables =
+      f.sformals @ f.slocals @ List.map duplicated_variable f.sformals @ reachable_from_args ctx args
+    in
+    let new_state = D.remove_terms_not_containing_variables reachable_variables state_with_duplicated_vars in
+    if M.tracing then M.trace "wrpointer-function" "ENTER2: result: %s\n" (D.show new_state);
+    [ctx.local, new_state]
+
+  (*ctx caller, t callee, ask callee, t_context_opt context vom callee -> C.t
+     expr funktionsaufruf*)
+  let combine_env ctx var_opt expr f exprs t_context_opt t (ask: Queries.ask) =
+    ctx.local
+
+  (*ctx.local is after combine_env, t callee*)
+  let combine_assign ctx var_opt expr f args t_context_opt t (ask: Queries.ask) =
+    let og_t = t in
+    (* assign function parameters to duplicated values *)
+    let arg_assigns = GobList.combine_short f.sformals args in
+    let state_with_assignments = List.fold_left (fun st (var, exp) -> assign_lval st (ask_of_ctx ctx) (Var (duplicated_variable var), NoOffset) exp) ctx.local arg_assigns in
+    if M.tracing then M.trace "wrpointer-function" "COMBINE_ASSIGN0: state_with_assignments: %s\n" (D.show state_with_assignments);
+    (*remove all variables that were tainted by the function*)
+    let tainted = (* find out the tainted variables from startState *)
+      ask.f (MayPointTo (MayBeEqual.return_lval (dummyFunDec.svar.vtype)))
+    in
+    if M.tracing then M.trace "wrpointer-tainted" "combine_env: %a\n" MayBeEqual.AD.pretty tainted;
+    let local = D.remove_tainted_terms ask tainted state_with_assignments in
+    let t = D.meet local t in
+    if M.tracing then M.trace "wrpointer-function" "COMBINE_ASSIGN1: var_opt: %a; local_state: %s; t_state: %s; meeting everything: %s\n" d_lval (BatOption.default (Var (MayBeEqual.dummy_varinfo (TVoid[])), NoOffset) var_opt) (D.show ctx.local) (D.show og_t) (D.show t);
+    let t = match var_opt with
+      | None -> t
+      | Some var -> assign_lval t ask var (MayBeEqual.return_lval (typeOfLval var))
+    in
+    if M.tracing then M.trace "wrpointer-function" "COMBINE_ASSIGN2: assigning return value: %s\n" (D.show_all t);
+    let local_vars = f.sformals @ f.slocals in
+    let duplicated_vars = List.map duplicated_variable f.sformals in
+    let t =
+      D.remove_terms_containing_variables (MayBeEqual.return_varinfo (TVoid [])::local_vars @ duplicated_vars) t
+    in if M.tracing then M.trace "wrpointer-function" "COMBINE_ASSIGN3: result: %s\n" (D.show t); t
+
+  let startstate v = D.top ()
+  let threadenter ctx ~multiple lval f args = [D.top ()]
+  let threadspawn ctx ~multiple lval f args fctx = D.top()
+  let exitstate  v = D.top ()
+
+end
+
+let _ =
+  MCP.register_analysis ~dep:["startState"; "taintPartialContexts"] (module Spec : MCPSpec)
```

Comment:
Is `taintPartialContexts` actually used here?

Reply:
I use it in `combine_env` for the query `MayBeTainted`